src/core/nodeOps.ts

```diff
@@ -115,71 +115,90 @@ export const nodeOps: RendererOptions<TresObject, TresObject> = {
+        throw 'could not find tres__deregisterBlockingObjectAtPointerEventHandler on scene\'s userData'
 
-        scene?.userData.tres__deregisterBlockingObjectAtPointerEventHandler?.(object as Object3D)
+      scene?.userData.tres__deregisterBlockingObjectAtPointerEventHandler?.(object as Object3D)
 
-        if (!deregisterAtPointerEventHandler)
-          throw 'could not find tres__deregisterAtPointerEventHandler on scene\'s userData'
+      if (!deregisterAtPointerEventHandler)
+        throw 'could not find tres__deregisterAtPointerEventHandler on scene\'s userData'
 
-        if (
-          object?.onClick
+      if (
+        object?.onClick
           || object?.onPointerMove
           || object?.onPointerEnter
           || object?.onPointerLeave
-        )
-          deregisterAtPointerEventHandler?.(object as Object3D)
-      }
+      )
+        deregisterAtPointerEventHandler?.(object as Object3D)
+    }
 
-      const deregisterCameraIfRequired = (object: Object3D) => {
-        const deregisterCamera = scene?.userData.tres__deregisterCamera
+    const deregisterCameraIfRequired = (object: Object3D) => {
+      const deregisterCamera = scene?.userData.tres__deregisterCamera
 
-        if (!deregisterCamera)
-          throw 'could not find tres__deregisterCamera on scene\'s userData'
+      if (!deregisterCamera)
+        throw 'could not find tres__deregisterCamera on scene\'s userData'
 
-        if ((object as Camera).isCamera)
-          deregisterCamera?.(object as Camera)
-      }
+      if ((object as Camera).isCamera)
+        deregisterCamera?.(object as Camera)
+    }
 
+    // Since remove is only called on the node being removed and not on child nodes.
+    // We need to recursively remove objects.
+    // In order for an object to be able to dispose it has to have
+    //   - a dispose method,
+    //   - it cannot be a <primitive object={...} />
+    //   - it cannot be a THREE.Scene, because three has broken it's own api
+    //
+    const isPrimitive = node?.primitive
+    const shouldDispose = !isPrimitive && (node.dispose === undefined ? node.dispose !== null : false)
+
+    // Add code to detach 
+    if (node?.isObject3D && node.parent?.isObject3D) {
+      node.parent.remove(node)
       node.removeFromParent?.()
-      object3D.traverse((child: Object3D) => {
-        disposeMaterialsAndGeometries(child)
-        deregisterCameraIfRequired(child)
-        deregisterAtPointerEventHandlerIfRequired?.(child as TresObject)
-      })
-
-      disposeMaterialsAndGeometries(object3D)
-      deregisterCameraIfRequired(object3D)
-      deregisterAtPointerEventHandlerIfRequired?.(object3D as TresObject)
     }
 
-    node.dispose?.()
+    // Only dispose if it's not a primitive and it has a dispose method,
+    //  if its a primitive, disposal needs to be handled by the user.
+    if (!isPrimitive && node?.isObject3D) {
+      node.children.forEach((child: TresObject) => nodeOps.remove(child))
```

Comment:
Shouldn't this always happen? Even if the node is a primitive or an Object3D the children should be disposed properly right?

Reply:
@Tinoooo 

I think the idea is that a primitive's `:object` is created by the user in `<script setup>`. If Tres cleans it up, it risks breaking the user's setup. 

E.g., the user might have plans to reuse the `:object` after the `<primitive />` has been removed.

[Fwiw, R3F seems to follow the same rules.
](https://github.com/pmndrs/react-three-fiber/blob/0ef66a1d23bf16ecd457dde92b0517ceec9861c5/packages/fiber/src/core/renderer.ts#L207)

Reply:
Yep basically what @andretchen0 stated.

Reply:
Got ya 👍